src/searcher/glue.rs

```diff
@@ -0,0 +1,262 @@
+    }
+
+    pub fn push(
+        &mut self,
+        match_with_absolute_offsets: Range,
+        current_beginning_of_line_offset: usize,
+    ) {
+        if self.reference_beginning_of_line_offset.is_none() {
+            self.reference_beginning_of_line_offset = Some(current_beginning_of_line_offset);
+        }
+        self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+            .push(Range::new(
+                match_with_absolute_offsets.start()
+                    - self.reference_beginning_of_line_offset.unwrap(),
+                match_with_absolute_offsets.end()
+                    - self.reference_beginning_of_line_offset.unwrap(),
+            ));
+    }
+}
+
+impl AsRef<[Range]> for AccumulatedExactMatches {
+    fn as_ref(&self) -> &[Range] {
+        &self.matches_with_offsets_relative_to_reference_beginning_of_line_offset
+    }
+}
+
+#[derive(Debug)]
+pub struct MultiLine<'s, S> {
+    config: &'s Config,
+    core: Core<'s, S>,
+    slice: &'s [u8],
+    last_match: Option<Range>,
+    accumulated_exact_matches: AccumulatedExactMatches,
+}
+
+impl<'s, S: Sink> MultiLine<'s, S> {
+    pub fn new(
+        searcher: &'s Searcher,
+        query_context: QueryContext,
+        slice: &'s [u8],
+        write_to: S,
+    ) -> MultiLine<'s, S> {
+        MultiLine {
+            config: &searcher.config,
+            core: Core::new(searcher, query_context, write_to),
+            slice,
+            last_match: None,
+            accumulated_exact_matches: Default::default(),
+        }
+    }
+
+    pub fn run(mut self) -> Result<(), S::Error> {
+        if self.core.begin()? {
+            let mut keepgoing = true;
+            let mut query_cursor = QueryCursor::new();
+            let tree = get_parser(self.core.query_context().language)
+                .parse(self.slice, None)
+                .unwrap();
+            let filter = get_loaded_filter(
+                self.core.query_context().filter_library_path.as_deref(),
+                self.core.query_context().filter_arg.as_deref(),
+            );
+            let query = self.core.query_context().query.clone();
+            let capture_index = self.core.query_context().capture_index;
+            let mut matches = query_cursor
+                .captures(&query, tree.root_node(), self.slice)
+                .filter_map(|(match_, found_capture_index)| {
+                    let found_capture_index = found_capture_index as u32;
+                    if found_capture_index != capture_index {
+                        return None;
+                    }
+                    let mut nodes_for_this_capture = match_.nodes_for_capture_index(capture_index);
+                    let single_captured_node = nodes_for_this_capture.next().unwrap();
+                    assert!(
+                        nodes_for_this_capture.next().is_none(),
+                        "I guess .captures() always wraps up the single capture like this?"
```

Comment:
This is worth sanity-checking, I found it strange that the "captures-oriented" query API had the same `Iterator::Item` type as the "matches-oriented" one